src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -38,67 +39,136 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                appendShouldProceedLocal = _factory.Local(outLocal);
+
+                AddPlaceholderReplacement(trailingParameter, appendShouldProceedLocal);
+            }
+
+            var handlerConstructionAssignment = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression(construction));
 
             var usesBoolReturn = data.UsesBoolReturns;
-            var builderPatternExpressions = ArrayBuilder<BoundExpression>.GetInstance(usesBoolReturn ? 2 : node.Parts.Length);
-            builderPatternExpressions.Add(builderConstruction);
-            BoundExpression? appendBinaryExpression = null;
+            var resultExpressions = ArrayBuilder<BoundExpression>.GetInstance(node.Parts.Length + 1);
             foreach (var currentPart in node.Parts)
             {
                 var appendCall = (BoundCall)currentPart;
                 Debug.Assert(usesBoolReturn == (appendCall.Method.ReturnType.SpecialType == SpecialType.System_Boolean));
 
-                var rewrittenArgs = VisitList(appendCall.Arguments);
-                var rewrittenAppendCall = MakeCall(
+                // The append call itself could have an interpolated string conversion that uses the builder local as the receiver
+                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
+                // preserved, but shouldn't affect any subsequent calls.
+                BoundExpression? appendReceiver = builderTemp;
+                Debug.Assert(appendCall.Method.RequiresInstanceReceiver);
+                var argRefKindsOpt = appendCall.ArgumentRefKindsOpt;
+
+                var rewrittenArguments = VisitArguments(
+                    appendCall.Syntax,
+                    appendCall.Arguments,
+                    appendCall.Method,
+                    appendCall.ArgsToParamsOpt,
+                    argRefKindsOpt,
+                    ref appendReceiver,
+                    out ArrayBuilder<LocalSymbol>? temps);
+
+                var rewrittenAppendCall = MakeArgumentsAndCall(
                     appendCall.Syntax,
-                    builderTemp,
+                    appendReceiver,
                     appendCall.Method,
-                    rewrittenArgs,
+                    rewrittenArguments,
                     appendCall.ArgumentRefKindsOpt,
                     appendCall.Expanded,
                     appendCall.InvokedAsExtensionMethod,
                     appendCall.ArgsToParamsOpt,
                     appendCall.ResultKind,
                     appendCall.Type,
+                    temps,
                     appendCall);
 
-                Debug.Assert(usesBoolReturn == (rewrittenAppendCall.Type!.SpecialType == SpecialType.System_Boolean));
-                if (usesBoolReturn)
+                Debug.Assert(usesBoolReturn == (appendCall.Type!.SpecialType == SpecialType.System_Boolean));
```

Comment:
This assertion is duplicate from one above (line 79). I  think it made more sense before (checking `rewrittenAppendCall` instead)